reference/src/representation/enums.md

```diff
@@ -0,0 +1,348 @@
+
+## Categories of enums
+
+**Empty enums.** Enums with no variants can never be instantiated and
+are equivalent to the `!` type. They do not accept any `#[repr]`
+annotations.
+
+**Fieldless enums.** The simplest form of enum is one where none of
+the variants have any fields:
+
+```rust
+enum SomeEnum {
+  Variant1,
+  Variant2,
+  Variant3,
+}
+```
+
+Such enums correspond quite closely with enums in the C language
+(though there are important differences as well). Presuming that they
+have more than one variant, these sorts of enums are always
+represented as a simple integer, though the size will vary.
+
+Fieldless enums may also specify the value of their discriminants
+explicitly:
+
+```rust
+enum SomeEnum {
+  Variant22 = 22,
+  Variant44 = 44,
+  Variant45,
+}
+```
+
+As in C, discriminant values that are not specified are defined as
+either 0 (for the first variant) or as one more than the prior
+variant.
+
+**Data-carrying enums.** Enums whose variants have fields are called
+"data-carrying" enums. Note that for the purposes of this definition,
+it is not relevant whether those fields are zero-sized. Therefore this
+enum is considered "data-carrying":
+
+```rust
+enum Foo {
+  Bar(()),
+  Baz,
+}
+```
+
+## repr annotations accepted on enums
+
+In general, enums may be annotation using the following `#[repr]` tags:
+
+- A specific integer type (called `Int` as a shorthand below):
+  - `#[repr(u8)]`
+  - `#[repr(u16)]`
+  - `#[repr(u32)]`
+  - `#[repr(u64)]`
+  - `#[repr(i8)]`
+  - `#[repr(i16)]`
+  - `#[repr(i32)]`
+  - `#[repr(i64)]`
+- C-compatible layout:
+  - `#[repr(C)]`
+- C-compatible layout with a specified discriminant size:
+  - `#[repr(C, u8)]`
+  - `#[repr(C, u16)]`
+  - etc
+
+Note that manually specifying the alignment using `#[repr(align)]` is
+not permitted on an enum.
+
+The set of repr annotations accepted by an enum depends on its category,
+as defined above:
+
+- Empty enums: no repr annotations are permitted.
+- Fieldless enums: `#[repr(Int)]`-style and `#[repr(C)]` annotations are permitted, but `#[repr(C, Int)]` annotations are not.
+- Data-carrying enums: all repr annotations are permitted.
+
+## Enum layout rules
+
+The rules for enum layout vary depending on the category.
+
+### Layout of an empty enum
+
+An **empty enum** is an enum with no variants; empty enums can never
+be instantiated and are logically equivalent to the "never type"
+`!`. `#[repr]` annotations are not accepted on empty enums. Empty
+enums are guaranteed to have the same layout as `!` (zero size and
+alignment 1).
+
+### Layout of a fieldless enum
+
+If there is no `#[repr]` attached to a fieldless enum, the compiler
+will represent it using an integer of sufficient size to store the
+discriminants for all possible variants -- note that if there is only
+one variant, then 0 bits are required, so it is possible that the enum
+may have zero size. In the absence of a `#[repr]` annotation, the
```

Comment:
I think we should open an issue about whether we want to guarantee that fieldless enums with one variant have zero size or not (and reference that issue here).